lib/ansible/plugins/action/reboot.py

```diff
@@ -0,0 +1,266 @@
+
+    def get_system_uptime(self):
+        command_result = self._low_level_execute_command(self.DEFAULT_UPTIME_COMMAND, sudoable=self.DEFAULT_SUDOABLE)
+
+        # For single board computers, e.g., Raspberry Pi, that lack a real time clock and are using fake-hwclock
+        # launched by systemd, the update of utmp/wtmp is not done correctly.
+        # Fall back to using uptime -s for those systems.
+        # https://github.com/systemd/systemd/issues/6057
+        if '1970-01-01 00:00' in command_result['stdout']:
+            command_result = self._low_level_execute_command('uptime -s', sudoable=self.DEFAULT_SUDOABLE)
+
+        if command_result['rc'] != 0:
+            raise AnsibleError("%s: failed to get host uptime info, rc: %d, stdout: %s, stderr: %s"
+                               % (self._task.action, command_result.rc, to_native(command_result['stdout']), to_native(command_result['stderr'])))
+
+        return command_result['stdout'].strip()
+
+    def check_uptime(self, before_uptime):
+        display.vvv("%s: attempting to get system uptime" % self._task.action)
+        connect_timeout = self._task.args.get('connect_timeout', self._task.args.get('connect_timeout_sec', self.DEFAULT_CONNECT_TIMEOUT))
+
+        # override connection timeout from defaults to custom value
+        if connect_timeout:
+            try:
+                self._connection.set_option("connection_timeout", connect_timeout)
+                self._connection.reset()
+            except AttributeError:
+                display.warning("Connection plugin does not allow the connection timeout to be overridden")
+
+        # try and get uptime
+        try:
+            current_uptime = self.get_system_uptime()
+        except Exception as e:
+            raise e
+
+        # FreeBSD returns an empty string immediately before reboot so adding a length
+        # check to prevent prematurely assuming system has rebooted
+        if len(current_uptime) == 0 or current_uptime == before_uptime:
+            raise Exception("uptime has not changed")
+
+    def run_test_command(self, **kwargs):
+        test_command = self._task.args.get('test_command', self.DEFAULT_TEST_COMMAND)
+        display.vvv("%s: attempting post-reboot test command '%s'" % (self._task.action, test_command))
+        command_result = self._low_level_execute_command(test_command, sudoable=self.DEFAULT_SUDOABLE)
+
+        result = {}
+        if command_result['rc'] != 0:
+            result['failed'] = True
+            result['msg'] = 'test command failed: %s %s' % (to_native(command_result['stderr'], to_native(command_result['stdout'])))
+        else:
+            result['msg'] = to_native(command_result['stdout'])
+
+        return result
+
+    def do_until_success_or_timeout(self, action, reboot_timeout, before_uptime, action_desc, fail_sleep=1):
+        max_end_time = datetime.utcnow() + timedelta(seconds=reboot_timeout)
+
+        while datetime.utcnow() < max_end_time:
```

Comment:
Maybe implement exponential backoff?

Reply:
I'm not sure how to do that. Would love to learn!

Reply:
I like it!